pkg/controller/controller_test.go

```diff
@@ -4671,12 +5134,170 @@ func TestProvisionFromPVC(t *testing.T) {
+			clonePVName:      pvName,
+			volOpts:          generatePVCForProvisionFromXnsPVC(ns1, srcName, fakeSc1, "PersistentVolumeClaim", &srcNamespace, nil, requestedBytes, "", true),
+			expectFinalizers: true,
+			xnsEnabled:       true,
+			refGrant:         newRefGrant("refGrant1", srcNamespace, ns1, coreapiGrp, "PersistentVolumeClaim"),
+			expectedPVSpec: &pvSpec{
+				Name:          pvName,
+				ReclaimPolicy: v1.PersistentVolumeReclaimDelete,
+				AccessModes:   []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+				Capacity: v1.ResourceList{
+					v1.ResourceName(v1.ResourceStorage): bytesToQuantity(requestedBytes),
+				},
+				CSIPVS: &v1.CSIPersistentVolumeSource{
+					Driver:       "test-driver",
+					VolumeHandle: "test-volume-id",
+					FSType:       "ext4",
+					VolumeAttributes: map[string]string{
+						"storage.kubernetes.io/csiProvisionerIdentity": "test-provisioner",
+					},
+				},
+			},
+		},
+		"fail provision with apiGroup nil and unsupportedKind xns PersistentVolumeClaim PersitentVolumeClaim data source with refgrant when CrossNamespaceVolumeDataSource feature enabled": {
+			clonePVName: pvName,
+			volOpts:     generatePVCForProvisionFromXnsPVC(ns1, srcName, fakeSc1, unsupportedKind, &srcNamespace, nil, requestedBytes, "", true),
+			expectErr:   true,
+			xnsEnabled:  true,
+			refGrant:    newRefGrant("refGrant1", srcNamespace, ns1, coreapiGrp, "PersistentVolumeClaim"),
+		},
+		"fail provision with xns PersistentVolumeClaim PersitentVolumeClaim data source with Bad refgrant when CrossNamespaceVolumeDataSource feature enabled": {
+			clonePVName: pvName,
+			volOpts:     generatePVCForProvisionFromXnsPVC(ns1, srcName, fakeSc1, "PersistentVolumeClaim", &srcNamespace, &coreapiGrp, requestedBytes, "", true),
+			expectErr:   true,
+			xnsEnabled:  true,
+			refGrant:    newRefGrant("refGrant1", srcNamespace, srcNamespace, coreapiGrp, "PersistentVolumeClaim"),
+		},
+		"fail provision with xns PersistentVolumeClaim PersitentVolumeClaim data source without refgrant when CrossNamespaceVolumeDataSource feature enabled": {
+			clonePVName: pvName,
+			volOpts:     generatePVCForProvisionFromXnsPVC(ns1, srcName, fakeSc1, "PersistentVolumeClaim", &srcNamespace, &coreapiGrp, requestedBytes, "", true),
+			expectErr:   true,
+			xnsEnabled:  true,
+		},
+		"fail provision with xns PersistentVolumeClaim PersitentVolumeClaim data source with refgrant when CrossNamespaceVolumeDataSource feature disabled": {
+			clonePVName: pvName,
+			volOpts:     generatePVCForProvisionFromXnsPVC(ns1, srcName, fakeSc1, "PersistentVolumeClaim", &srcNamespace, &coreapiGrp, requestedBytes, "", true),
+			expectErr:   true,
+			refGrant:    newRefGrant("refGrant1", srcNamespace, ns1, coreapiGrp, "PersistentVolumeClaim"),
+		},
 	}
 
 	for k, tc := range testcases {
 		tc := tc
 		t.Run(k, func(t *testing.T) {
-			t.Parallel()
+			if !tc.xnsEnabled {
+				t.Parallel()
```

Comment:
Why does the xns feature require serial tests?

Reply:
Because the test did not work with feature gate enabled when it was executed in parallel test.

Reply:
Hm that seems to imply there is some global shared state somewhere. I think it would be good to investigate more into that.

Reply:
`DefaultFeatureGate` defined in [utilfeature](https://github.com/kubernetes/apiserver/blob/master/pkg/util/feature/feature_gate.go#L32) seems a global value, therefore setting [value](https://github.com/kubernetes-csi/external-provisioner/blob/3bfb91586acdcce7c8cd1796ebe0ab2770cf880d/vendor/k8s.io/component-base/featuregate/testing/feature_gate.go#L35) during tests causes changes to a global value in each test.

Reply:
I added the link of the issue I created as a comment